design/cdk-bootstrap.md

```diff
@@ -0,0 +1,199 @@
+  This role will be assumable by any principal from the account(s) passed by the `trust` option.
+  
+  This role has the following permissions:
+  
+  ```json
+  {
+    "PolicyDocument": {
+      "Statement": [
+        {
+          "Action": [
+            "s3:GetObject*",    "s3:GetBucket*", "s3:List*",
+            "s3:DeleteObject*", "s3:PutObject*", "s3:Abort*"
+          ],
+          "Resource": [
+            "arn-of-the-S3-assets-bucket",
+            "arn-of-the-S3-assets-bucket/*"
+          ]
+        },
+        {
+          "Action": [
+            "ecr:PutImage",        "ecr:InitiateLayerUpload",
+            "ecr:UploadLayerPart", "ecr:CompleteLayerUpload"
+          ],
+          "Resource": "arn-of-the-docker-assets-ecr-repository"
+        }
+      ]
+    }
+  }
+  ```
+
+* An IAM role, called the **Action role**,
+  that will be assumed when executing the CloudFormation deployment actions
+  (CreateChangeSet and ExecuteChangeSet) in the continuous delivery pipeline.
+  It is also assumable by any principal from the account(s) passed by the `trust` option.
+  
+  This role has the following permissions:
+  
+  ```json
+  {
+    "PolicyDocument": {
+      "Statement": [
+        {
+          "Action": [
+            "cloudformation:CreateChangeSet",   "cloudformation:DeleteChangeSet",
+            "cloudformation:DescribeChangeSet", "cloudformation:DescribeStacks",
+            "cloudformation:ExecuteChangeSet"
+          ],
+          "Resource": "*"
+        },
+        {
+          "Action": "iam:PassRole",
+          "Resource": "arn-of-the-cfn-deployment-role"
+        }
+      ]
+    }
+  }
+  ```
+
+* An IAM role, called the **CFN deployment role**,
+  that will be used to perform the actual CFN stack deployment in the continuous delivery pipeline to this environment.
+  It is assumable *only* by the CloudFormation service principal
+  (this is for security reasons, as this role will have, necessarily, very wide permissions).
+  It will not have any inline policies,
+  but will instead have the Managed Policies attached that the user selected when prompted (see above).
+
+#### Physical resource names
+
+All of the above resources will be created with well-defined physical names -
+none of them will rely on automated CloudFormation naming.
+This is so that the other stages of CDK synthesis,
+like asset resolution, can rely on the concrete names
+(there is no reliable way to transfer this kind of information in an automated way across regions and/or accounts).
+This also allows for more fine-grained permissions -
+for instance, the continuous delivery pipeline needs to grant the **action role**
+permissions to read from the pipeline's bucket
+(to read the artifact that contains the CFN template to deploy);
+this way, it can add a well-defined name to the resource policy of the artifact bucket,
+instead of granting those permissions to all principals in the account.
+
```

Comment:
We should also indicate that the convention should be expressible using standard cloudformation functions (e.g. it can't use a hash on something).

Reply:
Can you expand on this? You've already mentioned that in the original PR, but I don't exactly understand why. Why not use hashing for the names?

Reply:
It's important to keep bootstrapping simple because this command line tool won't scale for more complex scenarios such as ogs with thousands of accounts and teams. I am exploring a few options for these use cases (StackSets, the approach taken by Amazon Deployment Framework, etc). Some of these solutions (namely StackSets) is limited to only allow deploying a simple CFN template. This means that we want the standard bootstrap template to be a simple and static template that can be hand-written and easily understood by users.

Philosophically, I am thinking about the bootstrapping stack like the "BIOS" of a computer. It is a very basic component, that leverages the minimum amount of abstractions and high level capabilities in order to allow the rest of the system to operate.

Reply:
Done.